pkg/server/server_sql.go

```diff
@@ -1115,6 +1123,24 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 	}, nil
 }
 
+func maybeCheckTenantAccess(ctx context.Context, codec keys.SQLCodec, db *kv.DB) error {
+	if codec.ForSystemTenant() {
+		// Skip check for system tenant and return early.
+		return nil
+	}
+	// Perform a simple read to the tenant in order to verify that this instance
+	// will be authorized to access the tenant keyspace.
+	err := db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		start := codec.TenantPrefix()
+		end := start.PrefixEnd()
+		if _, err := txn.Scan(ctx, start, end, 1); err != nil {
+			return err
+		}
+		return nil
+	})
+	return err
+}
```

Comment:
Oof this hurts in the context of https://github.com/cockroachdb/cockroach/issues/79043 

Is the deal that the `Init` just hangs because the retry gets stuck restarting internally? I think we can do some more plumbing to get that error propagated without needing to wait for the extra round-trip here. 

FWIW, I'd expect authorization errors to propagate to the rangefeed via this code here: https://github.com/cockroachdb/cockroach/blob/f7bb118d8298eabe16a8ce99146225f7133fe3a5/pkg/sql/sqlinstance/instancestorage/instancereader.go#L143-L145